lib/map-pool.js

```diff
@@ -0,0 +1,64 @@
+'use strict';
+
+const Pool = require("generic-pool");
+const mapnik = require('@carto/mapnik');
+
+// Create a new mapnik map object at `this.mapnik`. Requires that the mapfile
+// be localized with `this.localize()`. This can be called in repetition because
+// it won't recreate `this.mapnik`.
+module.exports = function createMapPool (uri, xml) {
+    const factory = {
+        create: mapCreateFunctor(uri, xml),
+        destroy : mapDestoryFunctor()
+    };
+    const options = {
+        max: uri.query.poolSize,
+        maxWaitingClients: uri.query.poolMaxWaitingClients
+    };
+
+    return Pool.createPool(factory, options);
+};
+
+function mapCreateFunctor (uri, xml) {
+
+    // This function should never reject ¯\_(ツ)_/¯
+    // see https://github.com/coopernurse/node-pool/issues/175
+    // see https://github.com/coopernurse/node-pool/issues/183
+    return function mapCreate () {
+        return new Promise((resolve) => {
+            try {
+                const { tileSize, bufferSize } = uri.query;
+                const map = new mapnik.Map(tileSize, tileSize);
+
+                map.bufferSize = bufferSize;
+
+                const mapOptions = {
+                    strict: uri.strict,
+                    base: uri.query.base
+                };
+
+                map.fromString(xml, mapOptions, (err, map) => {
+                    if (err) {
+                        return resolve(err)
+                    }
+
+                    return resolve(map);
+                });
+            } catch (err) {
+                return resolve(err);
+            }
+        });
+    };
+}
+
+function mapDestoryFunctor () {
```

Comment:
Typo: `mapDestroyFunction` and the usage

Reply:
https://en.wikipedia.org/wiki/Functor

It might the concept is not applied in the right way though.

Reply:
I didn't know the concept, thanks! 

And the other word? `destory` --> `destroy`

Reply:
hahaha! f***ing dyslexia!